### ISSUE TITLE
Losing connection to the etcdv3 leader makes the database unreachable 

### ISSUE THREAD

**Issue by EtienneM** [zalando/patroni#2552]
**Describe the bug**

We are a hosting company and host many PostgreSQL databases. We currently face an issue in our staging environment. We recently upgraded the etcd version, and this process led to all PostgreSQL databases using the etcd v3 API being unreachable. PostgreSQL databases still using the etcd v2 API are not impacted.

The issue is the following. Every time we restart the etcd leader, we get a loop in etcd logs with a lot of lines in an _infinite_ loop such as:

```
Feb 06 15:07:00 ip-10-0-0-XX1 etcd[3413019]: invalid auth token: <REDACTED>.291107319                                                                                                    
```

The `<REDACTED>` part varies. I assume it's one value for each database we host.

The PostgreSQL logs also contains some lines in an _infinite_ loop:

```
2023-02-06 15:35:39.534250861 +0100 CET [postgresql-2] 2023-02-06 14:35:39,533 ERROR: Request to server https://10.0.0.XX1:2381 failed: ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2023-02-06 15:35:39.543716381 +0100 CET [postgresql-2] 2023-02-06 14:35:39,543 ERROR: Request to server https://10.0.0.XX2:2381 failed: ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2023-02-06 15:35:39.645052796 +0100 CET [postgresql-2] 2023-02-06 14:35:39,644 ERROR: Request to server https://10.0.0.XX3:2381 failed: ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

And the PostgreSQL database becomes unreachable.

Then we tried to downgrade etcd from 3.4.23 to 3.4.21 and we fail to reproduce the issue, everything is working as expected. 

Our assumption is the following. The etcd 3.4.22 release contains the following PR: https://github.com/etcd-io/etcd/pull/14548. Before this PR, the returned error was always `ErrGRPCPermissionDenied` handled in Patroni here:

https://github.com/zalando/patroni/blob/1669a49b2d98869003350d15119e94d9f1758481/patroni/dcs/etcd3.py#L298-L305

After the etcd PR #14548, new error types were introduced, especially `ErrGRPCAuthOldRevision` which looks like the error that could happen in our cases and that is not handled at all in the etcd client from Patroni.

**To Reproduce**

1. Start etcd >= 3.4.22
2. Start a PostgreSQL database with Patroni configured to use etcd v3 API.
3. Restart the etcd leader.
4. The PostgreSQL database becomes unreachable

**Expected behavior**

I expect the PostgreSQL database to keep being reachable, even though the etcd leader restarts and a new leader is elected.

**Screenshots**

n/a

**Environment**
-   Patroni version: 2.1.4 and 2.1.7
-   PostgreSQL version: 14.6
-   DCS (and its version): etcd 3.4.23

**Patroni configuration file**
```
etcd3:
  hosts: ['10.0.0.XX1:2381','10.0.0.XX2:2381','10.0.0.XX3:2381']
  protocol: https
  username: <REDACTED>
  password: <REDACTED>

bootstrap:
  dcs:
    ttl: 60
    loop_wait: 20
    retry_timeout: 20
    maximum_lag_on_failover: 1048576
    postgresql:
      use_pg_rewind: true
      parameters:
        max_connections: 50
        max_wal_senders: 5
        wal_keep_size: 48MB
        max_locks_per_transaction: 64

  method: pgbackrest
  pgbackrest:
    command: /postgresql_bootstrap_replica --username=<REDACTED>
    keep_existing_recovery_conf: true
  post_bootstrap: /postgresql_post_bootstrap

restapi:
  certfile: /etc/ssl/databases/server.crt
  keyfile: /etc/ssl/databases/server.key
  connect_address: 192.168.100.5:8008

ctl:
  insecure: false
  cacert: /etc/ssl/databases/root.crt
  certfile: /etc/ssl/databases/server.crt

postgresql:
  data_dir: /var/lib/postgresql/14/postgresql0
  bin_dir: /usr/lib/postgresql/14/bin
  config_dir: /etc/postgresql/14/main
  pgpass: /tmp/pgpass0
  recovery_conf:
    restore_command: 'pgbackrest --config=/var/lib/postgresql/14/pgbackrest.conf --stanza=<REDACTED> archive-get %f "%p"'

  parameters:
    unix_socket_directories: '/tmp'
    ssl: 'on'
    ssl_cert_file: '/etc/ssl/databases/server.crt'
    ssl_key_file: '/etc/ssl/databases/server.key'
    ssl_ca_file: '/etc/ssl/databases/root.crt'
    ssl_ciphers: 'HIGH'
    wal_level: 'replica'
    archive_mode: 'on'
    archive_command: 'pgbackrest --config=/var/lib/postgresql/14/pgbackrest.conf --stanza=<REDACTED> archive-push %p'
  create_replica_methods:
    - pgbackrest
    - scalingo_basebackup
  pgbackrest:
    command: pgbackrest --stanza=<REDACTED> --delta restore
    keep_data: true
    no_params: true
  scalingo_basebackup:
    command: /postgresql_basebackup.sh
    retries: 2
    verbose: "true"
  pg_hba:
    - hostssl   replication replicator 192.168.100.0/24 md5
    - hostssl   replication replicator 127.0.0.1/32 md5
    - hostnossl replication all        0.0.0.0/0     reject
    - local     postgres    postgresql               peer

    - hostssl   all all 0.0.0.0/0 md5
    - hostnossl all all 0.0.0.0/0 reject
    - hostssl   all all ::0/0 md5
    - hostnossl all all ::0/0 reject


tags:
  nofailover: false
  noloadbalance: false
  clonefrom: false
  nosync: false
```

**patronictl show-config**
```
loop_wait: 20
maximum_lag_on_failover: 1048576
pause: false
postgresql:
  parameters:
    max_connections: 50
    max_locks_per_transaction: 64
    max_wal_senders: 5
    wal_keep_size: 48
  use_pg_rewind: true
retry_timeout: 20
ttl: 60
```

**Have you checked Patroni logs?**

During the incident we have in a loop:

```
2023-02-06 15:35:39.534250861 +0100 CET [postgresql-2] 2023-02-06 14:35:39,533 ERROR: Request to server https://10.0.0.XX1:2381 failed: ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2023-02-06 15:35:39.543716381 +0100 CET [postgresql-2] 2023-02-06 14:35:39,543 ERROR: Request to server https://10.0.0.XX2:2381 failed: ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2023-02-06 15:35:39.645052796 +0100 CET [postgresql-2] 2023-02-06 14:35:39,644 ERROR: Request to server https://10.0.0.XX3:2381 failed: ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

**Have you checked PostgreSQL logs?**
Please provide a snippet here

Nothing but the Patroni logs.

**Have you tried to use GitHub issue search?**
Maybe there is already a similar issue solved.

Yes we found the issue https://github.com/zalando/patroni/issues/2331 which contains the "invalid auth token" log. We tried with Patroni 2.1.7 to test the fix for this issue but it still does not work as expected.

**Additional context**

Let us know if you need any more information or if you want us to try something else. Our staging infrastructure is here to be broken again :)

**Comment by CyberDem0n** [comment]
Thank you @EtienneM for a very detailed big report, it is very very helpful!

It seems that they broke something in gRPC-gateway implementation. Now if we try making requests with invalid token it simply closing connection providing to meaningful HTTP response code:
**valid token**
```
$ curl -v -H 'authorization: EkhCgoIAMRLJIBBj.90' -X POST -d '{"key":"L3NlcnZpY2UvYmF0bWFuLw=="}' http://localhost:2379/v3/kv/range
*   Trying 127.0.0.1:2379...
* Connected to localhost (127.0.0.1) port 2379 (#0)
> POST /v3/kv/range HTTP/1.1
> Host: localhost:2379
> User-Agent: curl/7.81.0
> Accept: */*
> authorization: EkhCgoIAMRLJIBBj.90
> Content-Length: 34
> Content-Type: application/x-www-form-urlencoded
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Access-Control-Allow-Headers: accept, content-type, authorization
< Access-Control-Allow-Methods: POST, GET, OPTIONS, PUT, DELETE
< Access-Control-Allow-Origin: *
< Content-Type: application/json
< Grpc-Metadata-Content-Type: application/grpc
< Date: Tue, 07 Feb 2023 10:47:13 GMT
< Content-Length: 115
< 
* Connection #0 to host localhost left intact
{"header":{"cluster_id":"14841639068965178418","member_id":"10276657743932975437","revision":"76","raft_term":"7"}}
```

**invalid token**
```
$ curl -v -H 'authorization: EkhCgoIAMRLJIBBj.91' -X POST -d '{"key":"L3NlcnZpY2UvYmF0bWFuLw=="}' http://localhost:2379/v3/kv/range
*   Trying 127.0.0.1:2379...
* Connected to localhost (127.0.0.1) port 2379 (#0)
> POST /v3/kv/range HTTP/1.1
> Host: localhost:2379
> User-Agent: curl/7.81.0
> Accept: */*
> authorization: EkhCgoIAMRLJIBBj.91
> Content-Length: 34
> Content-Type: application/x-www-form-urlencoded
> 
* Empty reply from server
* Closing connection 0
curl: (52) Empty reply from server
```

While old etcd was giving **meaningful response**:
```
$ curl -v -H 'authorization: KzlMPwIMqjQEEJjs.12' -X POST -d '{"key":"L3NlcnZpY2UvYmF0bWFuLw=="}' http://localhost:2379/v3/kv/range
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:2379...
* Connected to localhost (127.0.0.1) port 2379 (#0)
> POST /v3/kv/range HTTP/1.1
> Host: localhost:2379
> User-Agent: curl/7.81.0
> Accept: */*
> authorization: KzlMPwIMqjQEEJjs.12
> Content-Length: 74
> Content-Type: application/x-www-form-urlencoded
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Content-Type: application/json
< Trailer: Grpc-Trailer-Content-Type
< Date: Tue, 07 Feb 2023 10:54:03 GMT
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
{"error":"etcdserver: invalid auth token","message":"etcdserver: invalid auth token","code":16}
```

For me it looks like regression in gRPC-gateway implementation in Etcd.

**Comment by Soulou** [comment]
Hi @CyberDem0n, could you do the same test with last `3.5` release also? (since you've the right local setup to do it) It would help us to know the strategy to adopt. Thanks a lot!

**Comment by CyberDem0n** [comment]
Update,
3.5.7, 3.5.6, and 3.4.22 are working without issues.
It seems that only 3.4.23 is affected.

**Comment by Soulou** [comment]
That’s great to know, thank you!LéoOn 7 Feb 2023, at 13:52, Alexander Kukushkin ***@***.***> wrote:﻿
Update,
3.5.7, 3.5.6, and 3.4.22 are working without issues.
It seems that only 3.4.23 is affected.

**Comment by CyberDem0n** [comment]
Opened an issue in Etcd repo: https://github.com/etcd-io/etcd/issues/15256

**Comment by CyberDem0n** [comment]
Update.
They confirmed the problem and it seems that fix will be included to 3.4.24

**Comment by EtienneM** [comment]
Thanks a lot @CyberDem0n for your quick and efficient answers :)
